### DISCUSSION TITLE
Shaders on PointLight2D to customize the light mask, possible ?

### DISCUSSION THREAD

**Discussion by MelPeslier** [godotengine/godot-proposals#9886] [2D]
**Is it possible to right a shader code to alter the texture the light is using ?**

I made a shader code to make some wave mask that I want to use on the light, so it mixes with the texture and i can have custom waves of light with it. I'm using lights because I want to use the sdf collisions, occluders and all the related lights features.

The problem is that it's not working, I wrote it into the fragment() and I also tested into the light(), but It doesn't work

Here is the shader code, and the result in the material preview, and the final result is the unaffected light in the scene (that have a fully suarred texture)

**Shader code :**
```
shader_type canvas_item;

group_uniforms Waves;
uniform float wave_radius = 0.5;
uniform float wave_outer_line_coefs = 0.05;
uniform float wave_current_radius;
uniform float wave_inner_radius;


float smooth_stroke(float _edge, float _aim, vec2 _width){
	float inner = smoothstep(_edge, _edge + _width.x, _aim);
	float outer = smoothstep(_edge - _width.y, _edge, _aim);
	return outer - inner;
}

void fragment() {
// * Center UV *
	vec2 uv = UV ;
	uv = uv * 2.0 - 1.0;

// * World mask * : Circle visible
	float circle_sdf = length(uv) / wave_radius ; // divide length(...) by  radius / screen.x
	float outer_edge = step(circle_sdf, wave_current_radius); // TODO : make smoothstep
	float inner_edge = smoothstep(circle_sdf, 0.0, wave_inner_radius);
	inner_edge = pow(inner_edge, 0.5);
	float inner_part = min(inner_edge, outer_edge);

// * Outer Line *
	float outer_line_mask = pow( smooth_stroke(circle_sdf, wave_current_radius, wave_outer_line_coefs * vec2(1.0, 0.25)), 2.0);
	float line_fade_out = smoothstep(1.0, 0.75, wave_current_radius);
	float wave_output = outer_line_mask * line_fade_out;
	
	COLOR.rgb = vec3( inner_part + wave_output ) ;

}
```
**Result  wanted :**
![image](https://github.com/godotengine/godot-proposals/assets/35856216/913ce584-e253-4a69-890e-4ed8c3605777)

**Unaffected scene :**
![image](https://github.com/godotengine/godot-proposals/assets/35856216/06e873b6-ec05-48ef-9d38-f4195b715ace)

Test using subviewports : https://github.com/godotengine/godot-proposals/discussions/9886#discussioncomment-9665642 ( just below first comment)

**Comment by AThousandShips** [comment]
Is the [`LIGHT_VERTEX`](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/canvas_item_shader.html#fragment-built-ins) `inout` not working for your case?

**Reply by MelPeslier** [reply]
Thanks for replying, but 

```
//void light() {
	// Called for every pixel for every light affecting the CanvasItem.
	// Uncomment to replace the default light processing function with this one.
//}
```

**Reply by AThousandShips** [reply]
What do you mean? I'm not talking about the light stage but the vertex stage, check the documentation I sent

**Reply by AThousandShips** [reply]
Have you tried using `LIGHT_VERTEX` and `SHADOW_VERTEX`?

**Reply by MelPeslier** [reply]
I'll try right now

**Reply by MelPeslier** [reply]
Left is the light, Right is the sprite
![image](https://github.com/godotengine/godot-proposals/assets/35856216/dbc900e0-9500-4534-bb4e-90756a051824)
COLOR.rgb = vec3( inner_part + wave_output ) ;
LIGHT_VERTEX.rgb = vec3( inner_part + wave_output );
It doen't do anything, Idon't understand what they give access too, I tried using them to output colors, but it's always 1 as the output for LIGHT_VERTEX.rg, and SHADOW, Do you have any hint on how to use them ?

**Reply by AThousandShips** [reply]
What do you mean it doesn't do anything? It looks like it gives a shadow in the top right? But I can't help you using it as I don't use it myself, but if you try it out and experiment I'm sure you can figure out how to use it

Or ask in the [forums](https://forum.godotengine.org/)

**Reply by MelPeslier** [reply]
No the shadow is gived with the occluder up here, doesn't have anything to do with it :)

**Reply by AThousandShips** [reply]
Then my bad that's the wrong tool for this, mistook your second image for desired results

You should probably use the `light()` function then to achieve relevant results

**Comment by MelPeslier** [comment]
I tried using viewports as the original texture, it does work with sprite2D, but still not with lights2D
Sprite using the viewport : 
![image](https://github.com/godotengine/godot-proposals/assets/35856216/14a839d0-1c01-4919-ad4b-50dd405960d9)

Light2D using the subviewport :
![image](https://github.com/godotengine/godot-proposals/assets/35856216/8642d908-1b9d-4c9d-8963-68d5c06e41e2)

Should light up the scene like this Light2D (this one using a gradient texture):
![image](https://github.com/godotengine/godot-proposals/assets/35856216/5bf4f64a-e31e-4d66-809d-bd1be0eadc3a)

How does is not use the viewport texture ?

**Reply by AThousandShips** [reply]
You can't use them for light textures, see:
* https://github.com/godotengine/godot/pull/88349
* https://github.com/godotengine/godot/issues/75868
